File: src/webterm/docker_watcher.py
"""Docker container event watcher for dynamic session management.

Watches Docker events and creates/removes terminal sessions for containers
with the 'webterm-command' label.
"""

from __future__ import annotations

import asyncio
import contextlib
import json
import logging
import os
from typing import TYPE_CHECKING, Callable

from .docker_stats import get_docker_socket_path

if TYPE_CHECKING:
    from .session_manager import SessionManager

log = logging.getLogger("webterm")

LABEL_NAME = "webterm-command"
THEME_LABEL = "webterm-theme"
# All labels that trigger container inclusion
WEBTERM_LABELS = (LABEL_NAME, THEME_LABEL)
AUTO_COMMAND_ENV = "WEBTERM_DOCKER_AUTO_COMMAND"
DEFAULT_COMMAND = "/bin/bash"
AUTO_COMMAND_SENTINEL = "__docker_exec__"


def _has_webterm_label(attributes: dict) -> bool:
    """Check if a container has any webterm label."""
    return any(label in attributes for label in WEBTERM_LABELS)


def _get_auto_command() -> str:
    return os.environ.get(AUTO_COMMAND_ENV, DEFAULT_COMMAND)


def _is_auto_label(value: str | None) -> bool:
    if value is None:
        return True
    stripped = value.strip()
    return stripped == "" or stripped.lower() == "auto"


class DockerWatcher:
    """Watch Docker events and manage terminal sessions dynamically."""

    def __init__(
        self,
        session_manager: SessionManager,
        on_container_added: Callable[[str, str, str], None] | None = None,
        on_container_removed: Callable[[str], None] | None = None,
        socket_path: str | None = None,
    ) -> None:
        """Initialize Docker watcher.

        Args:
            session_manager: Session manager for adding/removing apps.
            on_container_added: Callback(slug, name, command) when container is added.
            on_container_removed: Callback(slug) when container is removed.
            socket_path: Docker socket path (default: /var/run/docker.sock).
        """
        self._session_manager = session_manager
        self._on_container_added = on_container_added
        self._on_container_removed = on_container_removed
        self._socket_path = socket_path or get_docker_socket_path()
        self._running = False
        self._task: asyncio.Task | None = None
        # Track containers we're managing: slug -> container_id
        self._managed_containers: dict[str, str] = {}

    async def _docker_request(self, method: str, path: str) -> tuple[int, str]:
        """Make HTTP request to Docker socket.

        Returns:
            Tuple of (status_code, body).
        """
        reader, writer = await asyncio.open_unix_connection(self._socket_path)
        try:
            request = f"{method} {path} HTTP/1.1\r\nHost: localhost\r\n\r\n"
            writer.write(request.encode())
            await writer.drain()

            # Read status line
            status_line = await reader.readline()
            status_code = int(status_line.decode().split()[1])

            # Read headers
            content_length = 0
            chunked = False
            while True:
                line = await reader.readline()
                if line == b"\r\n":
                    break
                header = line.decode().lower()
                if header.startswith("content-length:"):
                    content_length = int(header.split(":")[1].strip())
                if "transfer-encoding: chunked" in header:
                    chunked = True

            # Read body
            if chunked:
                body_parts = []
                while True:
                    size_line = await reader.readline()
                    size = int(size_line.decode().strip(), 16)
                    if size == 0:
                        break
                    chunk = await reader.readexactly(size)
                    body_parts.append(chunk)
                    await reader.readline()  # trailing CRLF
                body = b"".join(body_parts).decode()
            elif content_length > 0:
                body = (await reader.readexactly(content_length)).decode()
            else:
                body = ""

            return status_code, body
        finally:
            writer.close()
            await writer.wait_closed()

    async def _get_labeled_containers(self) -> list[dict]:
        """Get all running containers with any webterm label.

        Queries for both webterm-command and webterm-theme labels,
        merging results and deduplicating by container ID.
        """
        seen_ids: set[str] = set()
        result: list[dict] = []

        for label in WEBTERM_LABELS:
            path = f'/containers/json?filters={{"label":["{label}"]}}'
            status, body = await self._docker_request("GET", path)
            if status != 200:
                log.error("Failed to list containers for label %s: %s", label, body)
                continue
            for container in json.loads(body):
                container_id = container.get("Id", "")
                if container_id and container_id not in seen_ids:
                    seen_ids.add(container_id)
                    result.append(container)

        return result

    def _get_container_command(self, container: dict) -> str:
        """Get command for container from label.

        If label is 'auto', returns default exec command.
        """
        labels = container.get("Labels", {})
        label_value = labels.get(LABEL_NAME)

        if _is_auto_label(label_value):
            return AUTO_COMMAND_SENTINEL
        return label_value

    def _get_container_theme(self, container: dict) -> str | None:
        labels = container.get("Labels", {})
        value = labels.get(THEME_LABEL)
        if isinstance(value, str) and value.strip():
            return value.strip()
        return None

    def _get_container_name(self, container: dict) -> str:
        """Get container name (without leading /)."""
        names = container.get("Names", [])
        if names:
            return names[0].lstrip("/")
        return container.get("Id", "unknown")[:12]

    def _container_to_slug(self, container: dict) -> str:
        """Convert container to URL slug."""
        return self._get_container_name(container).replace("_", "-").replace(".", "-")

    async def _add_container(self, container: dict) -> None:
        """Add a container as a terminal session."""
        slug = self._container_to_slug(container)
        name = self._get_container_name(container)
        command = self._get_container_command(container)
        theme = self._get_container_theme(container)
        container_id = container.get("Id", "")

        if slug in self._managed_containers:
            log.debug("Container %s already managed", name)
            return

        log.info("Adding container: %s (slug=%s, cmd=%s)", name, slug, command)
        self._managed_containers[slug] = container_id
        self._session_manager.add_app(name, command, slug, terminal=True, theme=theme)

        if self._on_container_added:
            self._on_container_added(slug, name, command)

    async def _remove_container(self, container_id: str) -> None:
        """Remove a container's terminal session."""
        # Find slug by container_id
        slug = None
        for s, cid in list(self._managed_containers.items()):
            if cid == container_id or cid.startswith(container_id):
                slug = s
                break

        if not slug:
            return

        log.info("Removing container: %s", slug)
        del self._managed_containers[slug]

        # Remove from session manager's apps
        if slug in self._session_manager.apps_by_slug:
            app = self._session_manager.apps_by_slug.pop(slug)
            if app in self._session_manager.apps:
                self._session_manager.apps.remove(app)

        # Close any active session for this slug
        route_key = slug  # In our case, slug is used as route_key
        session = self._session_manager.get_session_by_route_key(route_key)
        if session:
            session_id = self._session_manager.routes.get(route_key)
            if session_id:
                await self._session_manager.close_session(session_id)

        if self._on_container_removed:
            self._on_container_removed(slug)

    async def _watch_events(self) -> None:
        """Watch Docker events stream."""
        filters = json.dumps({"event": ["start", "die"], "type": ["container"]})
        path = f"/events?filters={filters}"

        while self._running:
            try:
                reader, writer = await asyncio.open_unix_connection(self._socket_path)
                try:
                    request = f"GET {path} HTTP/1.1\r\nHost: localhost\r\n\r\n"
                    writer.write(request.encode())
                    await writer.drain()

                    # Skip HTTP headers
                    while True:
                        line = await reader.readline()
                        if line == b"\r\n":
                            break

                    # Read event stream (chunked encoding)
                    while self._running:
                        size_line = await reader.readline()
                        if not size_line:
                            break
                        try:
                            size = int(size_line.decode().strip(), 16)
                        except ValueError:
                            continue
                        if size == 0:
                            break

                        chunk = await reader.readexactly(size)
                        await reader.readline()  # trailing CRLF

                        try:
                            event = json.loads(chunk.decode())
                            await self._handle_event(event)
                        except json.JSONDecodeError:
                            continue
                finally:
                    writer.close()
                    await writer.wait_closed()
            except Exception as e:
                if self._running:
                    log.warning("Docker event stream error: %s, reconnecting...", e)
                    await asyncio.sleep(5)

    async def _handle_event(self, event: dict) -> None:
        """Handle a Docker event."""
        action = event.get("Action", "")
        actor = event.get("Actor", {})
        container_id = actor.get("ID", "")
        attributes = actor.get("Attributes", {})

        # Only handle containers with any webterm label
        if not _has_webterm_label(attributes):
            return

        if action == "start":
            # Get full container info
            status, body = await self._docker_request("GET", f"/containers/{container_id}/json")
            if status == 200:
                container_info = json.loads(body)
                # Convert to list format expected by _add_container
                container = {
                    "Id": container_id,
                    "Names": ["/" + container_info.get("Name", "").lstrip("/")],
                    "Labels": container_info.get("Config", {}).get("Labels", {}),
                }
                await self._add_container(container)
        elif action == "die":
            await self._remove_container(container_id)

    async def scan_existing(self) -> None:
        """Scan for existing labeled containers and add them."""
        containers = await self._get_labeled_containers()
        for container in containers:
            await self._add_container(container)
        log.info("Found %d existing containers with %s label", len(containers), LABEL_NAME)

    async def start(self) -> None:
        """Start watching Docker events."""
        if self._running:
            return

        self._running = True
        # First scan existing containers
        await self.scan_existing()
        # Then start watching for new events
        self._task = asyncio.create_task(self._watch_events())
        log.info("Docker watcher started")

    async def stop(self) -> None:
        """Stop watching Docker events."""
        self._running = False
        if self._task:
            self._task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._task
            self._task = None
        log.info("Docker watcher stopped")

File: webterm/docker_watcher.go
package webterm

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

const (
	WebtermLabelName  = "webterm-command"
	WebtermThemeLabel = "webterm-theme"
)

type DockerWatcher struct {
	sessionManager *SessionManager
	socketPath     string

	onContainerAdded   func(slug, name, command string)
	onContainerRemoved func(slug string)

	mu         sync.RWMutex
	managed    map[string]string
	running    bool
	cancel     context.CancelFunc
	waitDone   chan struct{}
	httpClient *http.Client
}

func NewDockerWatcher(
	sessionManager *SessionManager,
	socketPath string,
	onAdded func(slug, name, command string),
	onRemoved func(slug string),
) *DockerWatcher {
	if socketPath == "" {
		socketPath = DockerSocketPath()
	}
	return &DockerWatcher{
		sessionManager:     sessionManager,
		socketPath:         socketPath,
		onContainerAdded:   onAdded,
		onContainerRemoved: onRemoved,
		managed:            map[string]string{},
		waitDone:           make(chan struct{}),
		httpClient:         newUnixHTTPClient(socketPath, 0),
	}
}

func hasWebtermLabel(labels map[string]string) bool {
	_, hasCommand := labels[WebtermLabelName]
	_, hasTheme := labels[WebtermThemeLabel]
	return hasCommand || hasTheme
}

func isAutoLabel(value string) bool {
	if strings.TrimSpace(value) == "" {
		return true
	}
	return strings.EqualFold(strings.TrimSpace(value), "auto")
}

func (w *DockerWatcher) getContainerCommand(container map[string]any) string {
	labels := toStringMap(container["Labels"])
	value := labels[WebtermLabelName]
	if isAutoLabel(value) {
		return AutoCommandSentinel
	}
	return value
}

func (w *DockerWatcher) getContainerTheme(container map[string]any) string {
	labels := toStringMap(container["Labels"])
	return strings.TrimSpace(labels[WebtermThemeLabel])
}

func (w *DockerWatcher) getContainerName(container map[string]any) string {
	names := toStringSlice(container["Names"])
	if len(names) > 0 {
		return strings.TrimPrefix(names[0], "/")
	}
	id := asString(container["Id"])
	if len(id) > 12 {
		id = id[:12]
	}
	return id
}

func (w *DockerWatcher) containerToSlug(container map[string]any) string {
	name := w.getContainerName(container)
	return strings.NewReplacer("_", "-", ".", "-").Replace(name)
}

func (w *DockerWatcher) addContainer(container map[string]any) {
	slug := w.containerToSlug(container)
	name := w.getContainerName(container)
	command := w.getContainerCommand(container)
	theme := w.getContainerTheme(container)
	containerID := asString(container["Id"])

	w.mu.Lock()
	if _, exists := w.managed[slug]; exists {
		w.mu.Unlock()
		return
	}
	w.managed[slug] = containerID
	w.mu.Unlock()

	w.sessionManager.AddApp(name, command, slug, true, theme)
	log.Printf("docker event: added container id=%s slug=%s name=%s", containerID, slug, name)
	if w.onContainerAdded != nil {
		w.onContainerAdded(slug, name, command)
	}
}

func (w *DockerWatcher) removeContainer(containerID string) {
	w.mu.Lock()
	slug := ""
	for s, id := range w.managed {
		if id == containerID || strings.HasPrefix(id, containerID) {
			slug = s
			delete(w.managed, s)
			break
		}
	}
	w.mu.Unlock()
	if slug == "" {
		return
	}

	if sessionID, ok := w.sessionManager.GetSessionIDByRouteKey(slug); ok {
		w.sessionManager.CloseSession(sessionID)
	}
	w.sessionManager.RemoveApp(slug)
	log.Printf("docker event: removed container id=%s slug=%s", containerID, slug)
	if w.onContainerRemoved != nil {
		w.onContainerRemoved(slug)
	}
}

func (w *DockerWatcher) listLabeledContainers() ([]map[string]any, error) {
	seen := map[string]bool{}
	containers := []map[string]any{}
	for _, label := range []string{WebtermLabelName, WebtermThemeLabel} {
		path := fmt.Sprintf(`/containers/json?filters={"label":["%s"]}`, label)
		status, body, err := unixJSONRequest(w.socketPath, http.MethodGet, path, nil)
		if err != nil || status != http.StatusOK {
			continue
		}
		var payload []map[string]any
		if err := json.Unmarshal(body, &payload); err != nil {
			continue
		}
		for _, container := range payload {
			id := asString(container["Id"])
			if id == "" || seen[id] {
				continue
			}
			seen[id] = true
			containers = append(containers, container)
		}
	}
	return containers, nil
}

func (w *DockerWatcher) handleEvent(event map[string]any) {
	action := asString(event["Action"])
	actor := toAnyMap(event["Actor"])
	containerID := asString(actor["ID"])
	if containerID == "" {
		return
	}
	switch action {
	case "start":
		log.Printf("docker event: action=start container=%s", containerID)
		path := fmt.Sprintf("/containers/%s/json", url.PathEscape(containerID))
		status, body, err := unixJSONRequest(w.socketPath, http.MethodGet, path, nil)
		if err != nil || status != http.StatusOK {
			log.Printf("docker event: start inspect failed container=%s status=%d err=%v", containerID, status, err)
			return
		}
		var detail map[string]any
		if err := json.Unmarshal(body, &detail); err != nil {
			log.Printf("docker event: start decode failed container=%s err=%v", containerID, err)
			return
		}
		config := toAnyMap(detail["Config"])
		labels := toStringMap(config["Labels"])
		if !hasWebtermLabel(labels) {
			return
		}
		container := map[string]any{
			"Id":     containerID,
			"Names":  []any{"/" + strings.TrimPrefix(asString(detail["Name"]), "/")},
			"Labels": config["Labels"],
		}
		w.addContainer(container)
	case "die":
		log.Printf("docker event: action=die container=%s", containerID)
		w.removeContainer(containerID)
	}
}

func (w *DockerWatcher) watchEvents(ctx context.Context, waitDone chan struct{}) {
	defer close(waitDone)
	filters := url.QueryEscape(`{"event":["start","die"],"type":["container"]}`)
	requestURL := "http://unix/events?filters=" + filters
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}
		req, _ := http.NewRequestWithContext(ctx, http.MethodGet, requestURL, nil)
		resp, err := w.httpClient.Do(req)
		if err != nil {
			log.Printf("docker event stream connect failed err=%v", err)
			select {
			case <-ctx.Done():
				return
			case <-time.After(5 * time.Second):
				continue
			}
		}
		log.Printf("docker event stream connected")
		decoder := json.NewDecoder(resp.Body)
		for {
			var event map[string]any
			if err := decoder.Decode(&event); err != nil {
				log.Printf("docker event stream decode error err=%v", err)
				break
			}
			w.handleEvent(event)
		}
		_ = resp.Body.Close()
		select {
		case <-ctx.Done():
			return
		case <-time.After(2 * time.Second):
		}
	}
}

func (w *DockerWatcher) ScanExisting() {
	containers, err := w.listLabeledContainers()
	if err != nil {
		log.Printf("docker scan failed err=%v", err)
		return
	}
	log.Printf("docker scan found %d labeled container(s)", len(containers))
	for _, container := range containers {
		w.addContainer(container)
	}
}

func (w *DockerWatcher) Start() {
	w.mu.Lock()
	if w.running {
		w.mu.Unlock()
		return
	}
	waitDone := make(chan struct{})
	ctx, cancel := context.WithCancel(context.Background())
	w.cancel = cancel
	w.waitDone = waitDone
	w.running = true
	w.mu.Unlock()
	log.Printf("docker watcher started")
	w.ScanExisting()
	go w.watchEvents(ctx, waitDone)
}

func (w *DockerWatcher) Stop() {
	w.mu.Lock()
	if !w.running {
		w.mu.Unlock()
		return
	}
	w.running = false
	cancel := w.cancel
	waitDone := w.waitDone
	w.mu.Unlock()
	if cancel != nil {
		cancel()
	}
	<-waitDone
	log.Printf("docker watcher stopped")
}
